fan_tools/red_metrics.py

```diff
@@ -0,0 +1,215 @@
+        for metric in self.simple_metrics:
+            _key = f'{key}:{metric}'
+            val = metrics.pop(_key)
+            result.append(f'{key}_{metric} {val.decode() if val else 0}')
+        duration_key = f'{key}:{Metrics.duration_value.value}:'
+        values = [float(v) for i, v in metrics.items() if str(i).startswith(duration_key)]
+        value = mean(values) if values else 0
+        result.append(f'{key}_duration_avg {value}')
+        return result
+
+    def get_metrics(self, key: str, mean_count: int = 10) -> List[str]:
+        metric_keys, expired_keys = self._get_metric_keys(key, mean_count=mean_count)
+        if expired_keys:
+            self.connection.delete(*expired_keys)
+        metrics = self._fetch_metrics(metric_keys)
+        return self._get_metrics(key=key, metrics=metrics)
+
+
+def red_metrics(name: str = None, storage: MetricsStorage = None):
+    """
+    Decorator for serve RED (Read, Error, Duration) metrics.
+    Usage:
+    >>> redis_storage = RedisMetricsStorage('redis://localhost:6379/1')
+    ... @red_metrics(name='some_code_metrics', storage=redis_storage):
+    ... def some_code_to_profile()
+    """
+
+    def get_metric_name(func, name: str = None) -> str:
+        return name or f'{func.__module__}.{func.__name__}'
+
+    @contextmanager
+    def prometheus_context(name: str):
+        start = time.time()
+        is_success = True
+        try:
+            yield
+        except:  # noqa 722
+            is_success = False
+            raise
+        finally:
+            if not storage:
+                return
+            execution_time = time.time() - start
+            if is_success:
+                storage.incr(f'{name}:{Metrics.requests_count.value}')
+            else:
+                storage.incr(f'{name}:{Metrics.errors_count.value}')
+            storage.sliding_window(f'{name}:{Metrics.duration_value.value}', execution_time)
+            storage.set(f'{name}:duration', execution_time)
+
+    def wrapper(func):
+        metric_name = get_metric_name(func, name)
+        if not asyncio.iscoroutinefunction(func):
+
+            @functools.wraps(func)
+            def wrapped(*args, **kwargs):
+                with prometheus_context(name=metric_name):
+                    return func(*args, **kwargs)
+
+        else:
+
+            @functools.wraps(func)
+            async def wrapped(*args, **kwargs):
+                with prometheus_context(name=metric_name):
+                    return await func(*args, **kwargs)
+
+        return wrapped
+
+    return wrapper
+
+
+@dataclass
+class ExporterMetricItem:
+    name: str
+    mean_count: int = 10
+
+
+class PrometheusExporter:
+    """
+    Decorator for providing RED (Read, Error, Duration) metrics.
+    Usage:
+    >>> redis_storage = RedisMetricsStorage('redis://localhost:6379/1')
+    ... exporter = PrometheusExporter(storage=redis_storage)
+    ... exporter.metrics([ExporterMetricItem(name='some_code_metrics', mean_count=5)])
```

Comment:
This declaration is way too long and requires us to register the same metric twice. 

I believe we could automatically register metrics in decorator and by default export everything with sane defaults.

The only thing that we possible can add to store different kind of metrics in the same redis instance - probably add some kind of prefix paramter

Reply:
because it would be tedious to repeat prefix parameter in every decorator, probably we can use default from environ, like `RED_PREFIX`